keps/sig-cloud-provider/1771-versioning-policy-for-external-cloud-providers/README.md

```diff
@@ -0,0 +1,121 @@
+- [ ] KEP approvers have approved the KEP status as `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Today we have many external (a.k.a out-of-tree) cloud providers, each responsible for managing releases for components
+such as cloud-controller-manger, but possibly others. Thus far there has been no standard on how providers should be
+semantically versioning their releases. Likewise, there has been no version skew policy indicating what versions of Kubernetes
+is supported by a version of a cloud-controller-manager (though it is assumed based on the vendored version of
+k8s.io/kubernetes). This KEP proposes to standardize all external cloud providers on a semantic versioning policy that ensures
+the vendored version of the cloud-controller-manager library is likely to match the version of the Kubernetes control plane.
+More concretely, releases for external cloud providers should track the major and minor versions of the Kubernetes version they
+intend to support. Patch releases are not required to match. For example, release v1.18.X of cloud-controller-manager for provider
+Foo should support Kubernetes clusters on v1.18.X.
+
+## Motivation
+
+* improve discovery of version supportability across Kubernetes and external cloud providers.
+* promote using the latest cloud-controller-manager library based on the Kubernetes version of the cluster
+* prevent wide version skew of Kubernetes control plane and cloud-controller-manager.
+* reduce chances of incompatibility issues that occur when a cloud-controller-manager vendoring a really
+old version of Kubernetes is running on a much newer version of Kubernetes.
+
+### Goals
+
+* standardize the semantic versioning of all external cloud providers
+* reduce chances of incompatibility issues due to version skew between cloud-controller-manager and Kubernetes.
+
+### Non-Goals
+
+* improving the build/release process for publishing arifacts like binaries and docker images.
+
+## Proposal
+
+Semantic versioning used by external cloud providers should track the major and minor releases of the supported Kubernetes
+version in that release. Patch releases are not required to match. For example, release v1.18.X of the cloud-controller-manager
```

Comment:
Ideally I think it should be more than just the CCM. It would be nice if the cloud provider release itself matched the K8s release. We should also think about the version of K8s repos which get pulled in. It would probably make sense that things like client-go should be V0.18.Y for any v1.18.X CCM/cloud provider release.

Reply:
I agree

Reply:
Once we cut the tie to k8s.io/kubernetes and only depend on k8s.io/cloud-provider, I will loop back and be more specific about that bit.